### ISSUE TITLE
interaction between clock and blocking VSync

### ISSUE THREAD

**Issue by hvdnsz** [pygame/pygame#2585]
**Environment:**

You can get some of this info from the text that pops up in the console when you run a pygame program.

- **Operating system** Linux Mint, Xfce4:
- **Python version** 3.9.4 :
- **SDL version** SDL 2.0.14:
- **PyGame version** 2.0.1:
- **Relevant hardware** Desktop PC:

**Current behavior:**

When I turn on VSync and add SCALED (only for get VSync) flag constantly moving sprites increasing their coordinates (I've noticed this when I printed out the sprite's coordinates) even the collisions are implemented. This happens after I moved the pygame window out of my screen and back. If I turn off VSync and SCALED flag, everything works fine.

**Expected behavior:**

When VSync and SCALED are enabled the moving sprites do not lose their control over their coordinates, even if they out of the (user's) screen.

**Screenshots**

![screenshot_1](https://user-images.githubusercontent.com/74798660/117550370-4495de00-b040-11eb-9258-dd1c8f2a8b44.png)
![screenshot_2](https://user-images.githubusercontent.com/74798660/117550371-452e7480-b040-11eb-8de2-41ef72d401ec.png)
![screenshot_3](https://user-images.githubusercontent.com/74798660/117550372-452e7480-b040-11eb-82ea-d047146312f2.png)
![screenshot_4](https://user-images.githubusercontent.com/74798660/117550373-45c70b00-b040-11eb-932d-a590ebb0567a.png)

**Steps to reproduce:**

1. Enable VSync and the SCALED flag.
2. Create a simple game with a constantly moving bouncing object.
3. Runt the game.
4. Watch the ball bouncing around the edges of the screen.
5. Grab the window and move it out of the screen as much as possible. Then drop it grab again and repeat it a few times.
6. Move the window back to centre and notice that, the bouncing sprite disappeared.

**Test code**

Here is the test program. Printing the coordinates is not implemented.
```
import pygame

pygame.init()

SCREEN_WIDTH = 800
SCREEN_HEIGHT = 600

flags = pygame.SCALED
screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT), flags, vsync=1)
fps = 60

# the bouncing sprite
CUBE_HEIGHT = 20
cube = pygame.Surface((CUBE_HEIGHT, CUBE_HEIGHT))
cube.fill((250, 0, 0))
pos_x = 400
pos_y = 300
speed = 30

clock = pygame.time.Clock()

running = True
while running:
    dt = clock.tick(fps) / 100
    # just for close button
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False

    # update
    pos_y += speed * dt
    if pos_y < 0 or pos_y + CUBE_HEIGHT > SCREEN_HEIGHT:
        speed *= -1

    # draw
    screen.fill((0, 0, 0))
    screen.blit(cube, (pos_x, pos_y))
    pygame.display.flip()

pygame.quit()

```

**Comment by robertpfeiffer** [comment]
Not a bug, but probably something we should document better?

**Comment by robertpfeiffer** [comment]
Should be documented in clock (fix your timestep, set a maximum to let the game lag rather than crash the simulation), as well as display.flip (can block when using vsync, vsync _does_ things)

**Comment by hvdnsz** [comment]
I don't really understand what you mean by "fix your timestep, set a maximum to let the game lag rather than crash the simulation". but it should be documented in display.update too.

**Comment by hvdnsz** [comment]
So, you suggest that, until the simulation update the physics only once before rendering the picture, the game loop should clamp the delta time (even it sometime produces lag) to avoid the crash.

**Comment by Starbuck5** [comment]
Here's a fixed version.

In the old version, if the cube is offscreen, it is lost, because every tick if flips direction. The old version works on the assumption that if the cube gets over the borders, it can get back out by reversing in one tick. If it can't, it is stuck in a back and forth.

In this version, I hard coded the direction of the speed based on whether it is above or below the screen, and also capped the delta time at 0.1 seconds, which means anything slower than 10 fps will be treated as 10 fps.

```
import pygame

pygame.init()

SCREEN_WIDTH = 800
SCREEN_HEIGHT = 600

flags = pygame.SCALED
screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT), flags, vsync=1)
fps = 60

# the bouncing sprite
CUBE_HEIGHT = 20
cube = pygame.Surface((CUBE_HEIGHT, CUBE_HEIGHT))
cube.fill((250, 0, 0))
pos_x = 400
pos_y = 300
cube_speed = 300
speed = cube_speed

clock = pygame.time.Clock()

running = True
while running:
    dt = clock.tick(fps) / 1000
    dt = min(dt, 0.1)
    
    # just for close button
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False

    # update
    pos_y += speed * dt
    if pos_y < 0:
        speed = cube_speed

    if pos_y + CUBE_HEIGHT > SCREEN_HEIGHT:
        speed = -cube_speed

    # draw
    screen.fill((0, 0, 0))
    screen.blit(cube, (pos_x, pos_y))
    pygame.display.flip()

pygame.quit()
```

**Comment by hvdnsz** [comment]
Thanks for the fixes, the answers and the article about timestep, but I'm only at beginner level in making 2D simulations/games and I just want to understand the behaviour of delta time first. Because of this, my level of experience is not high enough to improve this issue, so this issue could be closed unless there is something that I forget about (e.g need-docks tag).

**Comment by robertpfeiffer** [comment]
@hvdnsz if this is really annoying for you, I'll make a new issue so you won't get pinged while we are working on this. Otherwise, you'll see circa 4 more messages until we merge the better docs.
